File: src/visual-graph/builder/outermost-branch-under.ts
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import { isBranchScope } from "./is-branch-scope.js";

// Walk upward from scopeId and return the outermost branch scope strictly
// nested under branchId. Returns null when scopeId is branchId itself,
// when scopeId is not actually under branchId, or when there is no
// intervening branch scope. Used to discover the merge container immediately
// underneath a given outer branch (e.g. an if/else nested inside a switch
// case) so callers can recurse one level at a time instead of jumping
// straight to the deepest branch container.
export function outermostBranchUnder(
  branchId: string,
  scopeId: string,
  scopeMap: ReadonlyMap<string, SerializedScope>,
): string | null {
  if (scopeId === branchId) {
    return null;
  }
  let result: string | null = null;
  let cur: SerializedScope | null = scopeMap.get(scopeId) ?? null;
  while (cur && cur.id !== branchId) {
    if (isBranchScope(cur.id, scopeMap)) {
      result = cur.id;
    }
    if (!cur.upper) {
      return null;
    }
    cur = scopeMap.get(cur.upper) ?? null;
  }
  if (cur?.id !== branchId) {
    return null;
  }
  return result;
}

File: crates/unsnarl-visual-graph/src/builder/outermost_branch_under.rs
//! Mirrors `ts/src/visual-graph/builder/outermost-branch-under.ts`.

use std::collections::HashMap;

use unsnarl_ir::serialized::SerializedScope;

use super::is_branch_scope::is_branch_scope;

pub fn outermost_branch_under(
    branch_id: &str,
    scope_id: &str,
    scope_map: &HashMap<&str, &SerializedScope>,
) -> Option<String> {
    if scope_id == branch_id {
        return None;
    }
    let mut result: Option<String> = None;
    let mut cur = scope_map.get(scope_id).copied();
    while let Some(scope) = cur {
        if scope.id.value() == branch_id {
            break;
        }
        if is_branch_scope(scope.id.value(), scope_map) {
            result = Some(scope.id.value().to_string());
        }
        let upper = scope.upper.as_ref()?;
        cur = scope_map.get(upper.value()).copied();
    }
    let cur = cur?;
    if cur.id.value() != branch_id {
        return None;
    }
    result
}
